### PR TITLE
Address spec adherence feedbacks

### LINKED ISSUE #6335: Audit adherence to spec feedbacks
Opening this issue to track the adherence to spec feedbacks from the Quantstamp audit

- [x] 1. In beacon-chain/blockchain/process_attestation.go, the function onAttestation(), many parts of the specification for on_attestatation() delay consideration, rather than avoiding consideration. However, the function onAttestation() does not seem to ever delay - only returns (possibly with errors). Ensure that the intent of the specification is preserved.
- [x] 2. In beacon-chain/blockchain/process_attestation_helpers.go#157, it may not be entirely clear why the seed impact on attestation verification is omitted from the specification.
- [x] 3. Typo in beacon-chain/blockchain/process_block_helpers.go#45, “feature” -> “future”.
- [x] 4. Typo in beacon-chain/blockchain/process_block_helpers.go#444, “couldn’t received” -> “couldn’t be received”.
- [x] 5. In beacon-chain/blockchain/process_block_helpers.go#225, based on the definition of params.BeaconConfig().SafeSlotsToUpdateJustified, should return a false here instead of a true. We recommend clarifying the definition and/or this specific case.
- [x] 6. In beacon-chain/core/validators/validator.go, although the implementation of SlashValidator() conforms to the specification, it is unclear whether BeaconState.slashings should represent per-epoch sums of slashed effective balances (i.e., balances before slashing) or only deltas (i.e., the actual slashed balances).
- [x] 7. In beacon-chain/core/helpers/rewards_penalties.go#30, we check if total == 0, which relates to the max() function in the spec pseudocode. Although it should be semantically equivalent in practice, it may be best to strictly match the spec by changing the line to total < params.BeaconConfig().EffectiveBalanceIncrement.
- [x] 8. In beacon-chain/core/helpers/committee.go, it is unclear how the domain of certain calls to Seed() are determined. For example, in UpdateProposerIndicesInCache() on L344 DomainBeaconAttester is used, whereas in precomputeProposerIndices() on L371 DomainBeaconProposer is used.
- [x] 9. In beacon-chain/p2p/encoder/ssz.go, according to specification Encoding-dependent header: Req/Resp protocols using the ssz or ssz_snappy encoding strategies MUST encode the length of the raw SSZ bytes, encoded as an unsigned protobuf varint. On L34, the encoder contains a function Encode() that can violate the spec (as opposed to EncodeWithLength()). The function seems to be used in tests only. We suggest moving it to testing scope only and removing it from the encoder.
- [x] 10. Specification pseudocode does not always exactly match the ETH2 specification, including, but not limited to the following examples:
 1. In beacon-chain/core/helpers/validators.go in the pseudocode on L116, MIN_SEED_LOOKAHEAD is used instead of MAX_SEED_LOOKAHEAD.
 2. In beacon-chain/core/blocks/block_operations.go in the pseudocode for process_block_header, the function call signing_root(state.latest_block_header) is used instead of hash_tree_root(state.latest_block_header).
 3. In beacon-chain/core/helpers/committee.go, in the pseudocode for get_beacon_committee, we have index=epoch_offset, instead of index=(slot % SLOTS_PER_EPOCH) * committees_per_slot + index.
 4. In beacon-chain/core/state/transition.go in the pseudocode for state_transition, there are several missing snippets such as if validate_result: assert verify_block_signature(state, signed_block).
 5. In beacon-chain/core/validators/validator.go, after L113, the if-block is missing. There should be a line for whistleblower_index = proposer_index.
 6. The type Hash is often used instead of Bytes32.
 7. In beacon-chain/core/block_operations.go#626, it is unclear where the assertion assert data.index < get_committee_count_at_slot(state, data.slot) is enforced.
- [x] 11. In beacon-chain/core/block_operations.go#1038, it seems the depth parameter noted in the spec pseudocode is not used. On the other hand, the depth variable is write-only in the specification. We recommend aligning the specification and code with each other, perhaps, mixing in depth into the root. beacon chain/core/blocks/block_operations.go, the comments before ProcessRandao(), VerifyIndexedAttestation(),
- [x] 12. In beacon-chain/core/helpers/attestation.go, the comment before SlotSignature() does not match the specification for get_slot_signature().
- [x] 13. In beacon-chain/core/helpers/committee.go, the comment before BeaconCommitteeFromState() does not match the specification.
- [x] 14. In beacon-chain/core/helpers/rewards_penalties.go, the specification for get_total_balances() has a max() statement which does not appear to be faithfully implemented. In particular, if 0 < total < EffectiveBalanceIncrement, then the wrong value will be returned.
- [x] 15. Typo in beacon-chain/core/helpers/validators.go#116, “MIN” -> “MAX”.

### PR DESCRIPTION
**Pull request by terencechain**
**What type of PR is this?**
> Code health and code comments improvements

**What does this PR do? Why is it needed?**
Addressed spec adherence feedbacks from the audit

**Which issues(s) does this PR fix?**

Fixes #6335 

**Other notes for review**

### CODE REVIEW

File: beacon-chain/blockchain/process_attestation_helpers.go [hunk 1]
```diff
@@ -113,40 +112,7 @@ func (s *Service) verifyAttestation(ctx context.Context, baseState *stateTrie.Be
 	}
 	indexedAtt := attestationutil.ConvertToIndexed(ctx, a, committee)
 	if err := blocks.VerifyIndexedAttestation(ctx, baseState, indexedAtt); err != nil {
-		if err == helpers.ErrSigFailedToVerify {
```
Comment (terencechain) [inline beacon-chain/blockchain/process_attestation_helpers.go:116]: This code has been old back in v0.9. We haven't seen attestation fails to verify due to bad committee caching.
This was one of the spec feed back. Using that as a good opportunity to delete

File: beacon-chain/p2p/encoder/ssz.go [hunk 2]
```diff
@@ -43,21 +43,6 @@ func (e SszNetworkEncoder) doEncode(msg interface{}) ([]byte, error) {
 	return ssz.Marshal(msg)
 }
 
-// Encode the proto message to the io.Writer.
-func (e SszNetworkEncoder) Encode(w io.Writer, msg interface{}) (int, error) {
```
Comment (terencechain) [inline beacon-chain/p2p/encoder/ssz.go:47]: No longer relevant. We only used it in tests. Updated tests to use `EncodeGossip`

File: beacon-chain/blockchain/process_attestation.go [hunk 3]
```diff
@@ -20,7 +20,8 @@ import (
 var ErrTargetRootNotInDB = errors.New("target root does not exist in db")
 
 // onAttestation is called whenever an attestation is received, verifies the attestation is valid and saves
-/// it to the DB.
+// it to the DB. As a pure function, this does not delay attestation based on the spec outline. The delay is handled
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/process_attestation.go:23]: What does "pure function" mean in this context?

Comment (terencechain) [reply beacon-chain/blockchain/process_attestation.go:23]: Stateless, it doesn't "hold" attestations within the function. I'll update the comment to make it more clear

Comment (prestonvanloon) [reply beacon-chain/blockchain/process_attestation.go:23]: OK thanks. To me, a pure function has no dependencies
```go
func pureFn(a, b int) int {
  return a + b
}

func notPureFn(a, b int) int {
  return db.Lookup(a, b)
}
```

**Pull request merged**